pallets/connectors/src/inbound.rs

```diff
@@ -0,0 +1,270 @@
+	) -> DispatchResult {
+		// Retrieve investment details
+		let invest_id: T::TrancheCurrency = Self::derive_invest_id(pool_id, tranche_id)?;
+
+		// Determine post adjustment amount
+		let pre_amount = T::ForeignInvestment::redemption(&investor, invest_id.clone())?;
+		let post_amount = pre_amount.ensure_add(amount)?;
+
+		// Transfer tranche tokens from `DomainLocator` account of origination domain
+		T::Tokens::transfer(
+			invest_id.clone().into(),
+			&T::AccountConverter::convert(sending_domain),
+			&investor,
+			amount,
+			false,
+		)?;
+
+		T::ForeignInvestment::update_redemption(&investor, invest_id, post_amount)?;
+
+		Ok(())
+	}
+
+	/// Decreases an existing redemption order of the investor.
+	///
+	/// Transfers the decreased redemption amount from the investor account into
+	/// holdings of the `DomainLocator` account of origination domain of this
+	/// message.
+	///
+	/// Initiates a return `ExecutedDecreaseRedemption` message to refund the
+	/// decreased amount on the source domain. The dispatch of this message is
+	/// delayed until the execution of the redemption, e.g. at least until the
+	/// next epoch transition.
+	pub fn do_decrease_redemption(
+		pool_id: PoolIdOf<T>,
+		tranche_id: TrancheIdOf<T>,
+		investor: T::AccountId,
+		currency_index: GeneralCurrencyIndexOf<T>,
+		amount: <T as Config>::Balance,
+		sending_domain: DomainAddress,
+	) -> DispatchResult {
+		// Retrieve investment details
+		let invest_id: T::TrancheCurrency = Self::derive_invest_id(pool_id, tranche_id)?;
+		// NOTE: Required for relaying `ExecutedDecreaseRedemption` message
+		let _currency = Self::try_get_payment_currency(invest_id.clone(), currency_index)?;
+
+		// Determine post adjustment amount
+		let pre_amount = T::ForeignInvestment::redemption(&investor, invest_id.clone())?;
+		let post_amount = pre_amount.ensure_sub(amount)?;
+
+		T::ForeignInvestment::update_redemption(&investor, invest_id.clone(), post_amount)?;
+
+		// Transfer tranche tokens to `DomainLocator` account of
+		// origination domain
+		T::Tokens::transfer(
+			invest_id.into(),
+			&investor,
+			&T::AccountConverter::convert(sending_domain),
+			amount,
+			false,
+		)?;
+
+		// TODO(subsequent PR): Handle response `ExecutedDecreaseRedemption` message to
+		// source destination which should refund the decreased amount.
+		// Blocked by https://github.com/centrifuge/centrifuge-chain/pull/1363
+		// Should be handled by pallet-foreign-investments
+
+		Ok(())
+	}
+
+	/// Collect the results of a user's invest orders for the given investment
+	/// id. If any amounts are not fulfilled, they are directly appended to the
+	/// next active order for this investment.
+	pub fn do_collect_investment(
+		pool_id: PoolIdOf<T>,
+		tranche_id: TrancheIdOf<T>,
+		investor: T::AccountId,
+	) -> DispatchResult {
+		let invest_id: T::TrancheCurrency = Self::derive_invest_id(pool_id, tranche_id)?;
+
+		T::ForeignInvestment::collect_investment(investor, invest_id)
+	}
```

Comment:
TODO: Needs to trigger an `ExecutedCollectInvest` message.

Reply:
Can we handle this in a subsequent PR as proposed in the PR description? Some parameters of the the `Executed*` messages are not trivial to derive and thus should be tested properly as well.

Reply:
Of course. Just wanted to have todos in the codebase as there are for the other left ones

Reply:
[bd7ed80](https://github.com/centrifuge/centrifuge-chain/pull/1364/commits/bd7ed80f7c660b12db28f171943082f32af38657)